DataDog/datadog-agent
# [system-probe] Make port mapping code network namespace aware

### What does this PR do?

This set of changes makes the port mapping code in system-probe network namespace aware. Before we were only getting the ports for the system-probe process, which runs in the root network namespace. This has implications for containers that have listening ports, with traffic to them being marked incorrectly as outgoing (since system-probe is not aware of the listening ports), when it should be incoming, resulting in failed address resolutions.

Changes to port mapping code:
- ports are stored prefixed with network namespace's inode number
- new methods to do lookup by (network ns, port), instead of just (port)
- store pid as well as the port in bpf maps; the pid is then used to lookup the network namespace inode when adding to the port mappings in go code
- we check both the port and nat'd port when determining the connection direction

### Motivation

Failed resolutions in istio specifically. Traffic directed to an istio sidecar'd pod is redirected to 15006 inside the pod, and is marked as outgoing, when it should be marked as incoming. This is because port 15006 was not being picked up by the port mapping code (the initial dump).

### Additional Notes

### Describe your test plan
- make sure the agent and system-probe are not running
- run a local container: `docker run -d -p 8081:80 nginx`
- start the system-probe *only*
- run `curl localhost:8081`; wait a second or two
- run `curl localhost/connections --unix-socket /var/run/sysprobe/sysprobe.sock | jq '.conns[] | select(.laddr.port==80 and .direction=="incoming")'`; you should see some json printed out (exactly one object), if not this is a test failure
- run another local container: `docker run -d -p 8082:80 nginx`
- run `curl localhost:8081`
- run `curl localhost:8082`; wait a second or two
- run `curl localhost/connections --unix-socket /var/run/sysprobe/sysprobe.sock | jq '.conns[] | select(.laddr.port==80 and .direction=="incoming")'`; you should see some json printed out (exactly two objects this time, with different laddr.ip fields), if not this is a test failure
